Improve IPDK cli and ci build/run with buildx, multiarch, artefacts and push to registries

stolsma opened the pull request:
Implements large parts of #101

Expanded IPDK CLI commands and options with:
- Support `ipdk config [key]=[value]` for setting key=value in ~/.ipdk/ipdk.env
- Support `ipdk build --platform [docker buildx supported platforms]` for local and multiarch builds
- Support `ipdk build --tags [taglist] --labels [labellist] --push --export [filename]`
- Support `ipdk start --platform [docker buildx supported platform]`
- Support `ipdk tag [tag]` for local image builds
- Support `ipdk export [filename]` for local image builds
- Support `ipdk push [tag]` for local image builds
- check for support of `docker buildx` on host system

Example:
`ipdk build --no-cache --platform linux/arm64` builds a ARM64 based ipdk image if run on a linux/amd-64 based host system
`ipdk start --platform linux/arm64` starts a ARM64 based image on a linux/amd-64 based host system by using qemu emulation

Expanded the Github `makefile.yml` action file with:
- concurrency option to prevent running multiple build actions for the same PR or push.
- install of qemu runtimes
- install of buildx
- (for now disabled) login to dockerhub
- login to Github Container Registry (ghcr)
- create build metadata (tags and labels)
- use ipdk cli build command
- push the created multiarch container images to the registries when run from push
- upload created multiarch container images as artefacts when the action is run from PR
- Currently only building containers for one architecture platform (linux/amd-64)

Signed-off-by: stolsma <github@tolsma.net>

stolsma commented:
First stab at implementing multi platform docker images! Please review + need a lot to test and figure out...

mestery commented:
Won't this take a really long time to run because of the Qemu emulation?

stolsma commented:
> Won't this take a really long time to run because of the Qemu emulation?

Hahaha, yes, thats why I didn't use --platform linux/arm64 (for example) to start with ;-) but it fails anyway as buildx needs to be installed on the github actions ubuntu image....

mestery commented:
Looks like the arm64 requests are taking too long. I think I'll try to setup the OCI GitHub runner today to get that part working and we can do arm64 builds there.

stolsma commented:
Yeah, those aarch64 builds are taking too long, but what is worse is that some commands take more then the maximum intermediate response time of github so the runner thinks its dead... Hopefully the Arm runners will help here!!!

stolsma commented:
Added concurrency to prevent several checks for the same PR at the same time, the last push will be run by Github Actions. If more are running they are cancelled.

Added Artefact saving, tagging and prepared DockerHub saving and Github Container Registry saving.

Will let it rest until after the weekend! Please review!

mestery reviewed (commented):
Do you want to separate out the platform support from this PR? It seems as if you're making enhancements to the CI which are not related to the arm platform anymore. Also, I think we should just the OCI arm runners for arm builds. I am going to work on getting that up and running at some point today.

mestery commented on .github/workflows/makefile.yml:
```diff
@@ -15,31 +19,59 @@ jobs:
+    - name: Container build
+      run: |
+        ipdk build --no-cache --platform ${{ matrix.platform }}
+        ipdk tag ipdk-io/ipdk:${{ matrix.os }}-latest 
```
I think we should tag this with the commit SHA has as well.

stolsma replied:
Agree

stolsma replied:
Implemented... The images are tagged with several tags depending on github actions event. One of the default tags us :sha-<short sha>

mestery commented on .github/workflows/makefile.yml:
```diff
@@ -15,31 +19,59 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu2004, fedora33]
+        platform: [linux/amd64]
```
Since you have dockerx working now, can we add back in `linux/arm64` here as well?

stolsma replied:
@mestery No, first the Dockerhub or Guthub Container Registry needs to be enabled (and some script code in the ipdk.sh build function  needs to be adapted) because the standard local docker image store doesn't support multi arch images 😔 Buildx needs to push directly to one of those remote registries. One solution I found was talking about a local registry as action service, but I didn't have enough time today to test that. Will try that on Monday...

mestery replied:
OK. I pushed a change to try this out before you had replied, feel free to force push back over top of that whenever you want to get rid of it.

stolsma replied:
I think I solved creating the multiarch problem. But now we need to solve the crashing linux/arm64 build

stolsma replied:
We can add `linux/arm64` to the platform environment variable when the pr's for arm64 support are pushed I think. Also it would be nice if we can add a native arm64 builder host to buildx but that probably depends on your endeavor getting oci access....

rst0git commented on .github/workflows/makefile.yml:
```diff
@@ -15,31 +26,84 @@ jobs:
+          --export /tmp/${{ matrix.os }}.tar
+
+    - name: Upload build artifact
+      if: github.event_name == 'pull_request'
```
I am not sure if this is a good idea. Uploading Fedora 33 (1.06 GB) + Ubuntu 20.04 (862 MB) artifacts for a few days, every time a pull request has been created or updated, might result in the error message: [`Cannot upload artifacts. You may have reached your capacity.`](https://github.community/t/reached-artifact-limit/16860)

stolsma replied:
Thanks for mentioning this limit possibility!!!

I have checked the docs and everywhere it is stated that the (artifact) storage limit is for private repositories only (and all the people mentioning the problem are talking about private repositories). The other limit that I have found is api calls (total calls in x time) but we don't have that problem here as it is only one file (and as such a very limited number of calls). Also repeated updated pr's within the runtime if the previous pr push are canceled an as such not stored. 

But you are right, storing for more then one day is not needed so I will add `retention-days: 1` so the artifacts will be purged after 1 day.

rst0git replied:
@stolsma Thanks for looking into this. Specifying `retention-days` sounds good to me and I think you are right. I checked with the GitHub virtual support and the answer was:

> Actions usage is free for public repositories and self-hosted runners. For private repositories, each account receives a certain amount of free minutes and storage.

@mestery What do you think?

mestery replied:
One thing we could do is only push images on merges, which would limit how much storage we use and stop pushing images which we build as part of pull request testing. Another thing we could do is set a retention policy (see [here](https://github.com/actions/upload-artifact/issues/59#issuecomment-1026859091) for things in a reasonable manner.

mestery replied:
I think 1 day retention is fine, and only pushing on merge is also the right move, thanks for taking care of that!

stolsma commented:
Ok, I think everything is implemented to support several solutions of building IPDK containers. Support is available for base distribution selection, single/multi architecture image builds, tags & label adding, save to artefacts and push to registries. All those options also work for local builds and Github CI builds but with the caveat that the local docker image store only supports single architecture images! (that's a docker limitation!!!)

The three remaining questions are:
1. are we going to support multiarch images and how are we going to build them ? See also 3. 
2. how we are going to speed up build time and get the images smaller
    - my opinion is that we need to split the build in a development image build (with all extra tools installed) and a production image build used in CI (only with the libraries to get the IPDK programs running). Both can be made available on Dockerhub & GHCR. The development image can for example automatically be build and pushed to registries once a day and the production images build and pushed to registries on push to main (and for testing with PRs).
3. how are we going to integrate private build environments like OCI for ARM64.
   - Two options to select from: 
       1. use buildx builder nodes to add a OCI based docker buildx ARM64 environment and create multiarch images
       2. use a OCI based ARM64 runner and create single arch images

@mestery and others, what would you think????

mestery commented:
> Ok, I think everything is implemented to support several solutions of building IPDK containers. Support is available for base distribution selection, single/multi architecture image builds, tags & label adding, save to artefacts and push to registries. All those options also work for local builds and Github CI builds but with the caveat that the local docker image store only supports single architecture images! (that's a docker limitation!!!)
> 
> The three remaining questions are:
> 
> 1. are we going to support multiarch images and how are we going to build them ? See also 3.

I think we should, but I don't think we can do it using docker buildx with Qemu on the default runners GitHub Actions uses because it's too slow. Ideally, we'd use something like OCI and have our own native arm64 runners.

> 2. how we are going to speed up build time and get the images smaller
>    
>    * my opinion is that we need to split the build in a development image build (with all extra tools installed) and a production image build used in CI (only with the libraries to get the IPDK programs running). Both can be made available on Dockerhub & GHCR. The development image can for example automatically be build and pushed to registries once a day and the production images build and pushed to registries on push to main (and for testing with PRs).

My thinking is that IPDK should not be building P4-OVS each time, and that we should move to focus on the OVS repository and get proper CI which builds P4-OVS and stores it somewhere. The IPDK build will then pull the latest version of P4-OVS. This means the IPDK build would be really fast, and only people modifying P4-OVS would require the hour long build we currently have.

> 3. how are we going to integrate private build environments like OCI for ARM64.
>    
>    * Two options to select from:
>      
>      1. use buildx builder nodes to add a OCI based docker buildx ARM64 environment and create multiarch images
>      2. use a OCI based ARM64 runner and create single arch images
> 

I think we'd just integrate arm64 runners on OCI by making them runners.

> @mestery and others, what would you think????

stolsma commented:
> I think we'd just integrate arm64 runners on OCI by making them runners.

@mestery I will change the PR then to use a `os` and `platform` matrix to create (at this moment) a `Ubuntu18.04-amd64`, `Ubuntu20.04-amd-64` and `fedora33-amd64` container image. When OCI for arm64 is added as runner then arm64 can be added to the `platform` matrix and three `*-arm64` will also be created next to the previous three. Agreed?

> My thinking is that IPDK should not be building P4-OVS each time, and that we should move to focus on the OVS repository and get proper CI which builds P4-OVS and stores it somewhere. The IPDK build will then pull the latest version of P4-OVS. This means the IPDK build would be really fast, and only people modifying P4-OVS would require the hour long build we currently have.

Agree, I think we should split a "once a day build" for p4c as well. That shouldn't be a big problem as the (container) build examples already exist for Ubuntu and Fedora and can easily be merged with the IPDK build just like p4-ovs.

stolsma commented:
@mestery If the last run checks out ok it is ready for review and merge!!! Changed the action so OCI runner support can be added later by changing `matrix.runner` directive!

rst0git commented on build/scripts/ipdk-lib.sh:
```diff
@@ -13,6 +13,73 @@ print_banner() {
+		return 1
+	fi
+
+	docker_version="$(docker --version | cut -d' ' -f3 | tr -cd '0-9.')"
```
The following command might be more reliable way of getting the docker version:
```
 docker version --format '{{.Server.Version}}'
```

rst0git commented on build/scripts/ipdk-lib.sh:
```diff
@@ -13,6 +13,73 @@ print_banner() {
+		return 1
+	fi
+
+	docker_experimental="$(docker version | awk '/^ *Experimental:/ {print $2 ; exit}')"
```
```
docker version --format='{{.Server.Experimental}}'
```

mestery commented:
> @mestery If the last run checks out ok it is ready for review and merge!!! Changed the action so OCI runner support can be added later by changing `matrix.runner` directive!

This is great! I'll merge this one once it passes (and I fix @rst0git's comments), and then this afternoon I'll create an OCI account and try to get an arm64 runner working there!
